runtime/gc_base/StringTable.cpp

```diff
@@ -976,4 +975,58 @@ j9gc_internString(J9VMThread *vmThread, j9object_t sourceString)
 	return internedString;
 }
 
+typedef struct {
+	J9UTF8* utf;
+	j9object_t stringObject;
+} J9UTFCacheEntry;
+
+static UDATA
+utfCacheHashFn(void *key, void *userData)
+{
+	/* UTF8 pointers are 2-aligned, so shift the pointer for maximum entropy */
+	return ((UDATA)((J9UTFCacheEntry*)key)->utf) >> 1;
+}
+
+static UDATA
+utfCacheHashEqualFn(void *leftKey, void *rightKey, void *userData)
+{
+	return ((J9UTFCacheEntry*)leftKey)->utf == ((J9UTFCacheEntry*)rightKey)->utf;
+}
+
+j9object_t
+j9gc_createJavaLangStringFromUTF(J9VMThread *vmThread, J9UTF8 *utf, UDATA stringFlags)
+{
+	bool cacheString = J9_ARE_ANY_BITS_SET(stringFlags, J9_STR_INTERN);
```

Comment:
Do we need to distinguish between the `UTF8`->`J9_STR_INTERN String` and `UTF8`->`J9_STR_INTERN | J9_STR_XLAT String` ?  If we don't, we may incorrect return non-XLAT'd strings or vice versa

Reply:
Yes, this currently works only in the most specific case, which may include those flags. I'll have another look, and possibly delete the flags parameter if all the callers agree.